### PR TITLE
btrfs-progs: add offline conversion to remap tree

### PR DESCRIPTION
**Pull request by maharmstone**
Add an option `--convert-to-remap-tree` to btrfstune, which converts a filesystem to use the new remap tree feature. There's no `--convert-from-remap-tree` yet, as it's considerably more complicated: it implies undoing any remappings that have already taken place.

This does the following:

* Enables the block group tree and free space tree, if these aren't already set
* Sets the remap tree incompat flag
* Removes the data reloc tree
* Creates a METADATA_REMAP block group
* Creates an empty remap tree
* Extends the block group items to their new form

### CODE REVIEW

File: tune/convert-remap-tree.c [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+	root = btrfs_read_fs_root(fs_info, &key);
+	if (IS_ERR(root))
+		return PTR_ERR(root);
+
+	ret = btrfs_del_root(trans, fs_info->tree_root, &root->root_key);
+	if (ret)
+		return ret;
+
+	ret = btrfs_free_tree_block(trans, btrfs_root_id(root), root->node,
```
Comment (adam900710) [inline tune/convert-remap-tree.c]: Why not to use `btrfs_delete_and_free_root()`? As I find this part missing the eb freeing of that function.

And maybe some checks on the level of the tree just in case.
I know you have checked that no running balance before going here, but if the tree by somehow has a lot of inode items, it can still cause problems.

File: tune/convert-remap-tree.c [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	if (ret < 0)
+		return ret;
+
+	leaf = path.nodes[0];
+
+	while (true) {
+		if (path.slots[0] >= btrfs_header_nritems(path.nodes[0])) {
+			ret = btrfs_next_leaf(fs_info->block_group_root, &path);
+			if (ret)
+				break;
```
Comment (adam900710) [inline tune/convert-remap-tree.c]: Missing error handling of `ret < 0` case.
In fact this break only works for `ret > 0` case.

File: tune/convert-remap-tree.c [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	ret = btrfs_search_slot(trans, fs_info->block_group_root, &key, &path,
+				-1, 1);
+	if (ret < 0)
+		return ret;
+
+	leaf = path.nodes[0];
+
+	while (true) {
+		if (path.slots[0] >= btrfs_header_nritems(path.nodes[0])) {
+			ret = btrfs_next_leaf(fs_info->block_group_root, &path);
```
Comment (adam900710) [inline tune/convert-remap-tree.c]: Is `btrfS_next_leaf()` safe here?

It doesn't COW the next leaf IIRC.

Thus the safest way to do this is to search the tree with COW for every block group item.
(Which can be fetched from chunk mapping or block group cache)

**Comment by maharmstone** [comment]
Thank you Qu, all of that sounds correct.

I've pushed another version with your changes, would you mind having a look at that please? I also realized that it wasn't populating the remap root details in the superblock (there's no test for that yet!), so I've added another patch doing that

**Comment by adam900710** [comment]
The code looks good to me now. Although please add a basic test case, with at least a newly created empty btrfs to be converted to remap tree.
Preferably with some contents populated and converted to remap tree.

But in the long run, the single large transaction solution must be changed, or it will not be able to handle huge fses.

**Comment by maharmstone** [comment]
Thanks Qu, I've added a simple test

**Comment by adam900710** [comment]
Thanks a lot Mark, this version looks good to me now. Merged.
